File: menu.php
<?php
header('Content-Type: application/json');
require_once __DIR__ . '/utils.php';

$method = $_SERVER['REQUEST_METHOD'];
$rawPayload = null;
$payload = [];

if ($method !== 'GET') {
  $rawPayload = file_get_contents('php://input');
  $decoded = json_decode($rawPayload, true);
  if (is_array($decoded)) {
    $payload = $decoded;
  }
}

function normalize_menu_structure($menu) {
  if (!is_array($menu)) {
    $menu = [];
  }
  if (!isset($menu['drinks']) || !is_array($menu['drinks'])) {
    $menu['drinks'] = [];
  }
  if (!isset($menu['foods']) || !is_array($menu['foods'])) {
    $menu['foods'] = [];
  }
  $menu['drinks'] = array_values(array_filter(array_map('normalize_menu_item', $menu['drinks']), function ($item) {
    return $item['name'] !== '' && count($item['options']) > 0;
  }));
  $menu['foods'] = array_values(array_filter(array_map('normalize_menu_item', $menu['foods']), function ($item) {
    return $item['name'] !== '' && count($item['options']) > 0;
  }));
  return $menu;
}

function normalize_menu_item($item) {
  $name = '';
  $options = [];
  if (is_array($item)) {
    $name = trim((string)($item['name'] ?? ''));
    $optionsList = $item['options'] ?? [];
    if (is_array($optionsList)) {
      $unique = [];
      foreach ($optionsList as $option) {
        if (!is_string($option)) continue;
        $trimmed = trim($option);
        if ($trimmed === '' || in_array($trimmed, $unique, true)) continue;
        $unique[] = $trimmed;
      }
      $options = $unique;
    }
  }
  return ['name' => $name, 'options' => $options];
}

function persist_menu($menu) {
  if (!write_json_file(MENU_FILE, $menu)) {
    http_response_code(500);
    echo json_encode(["success" => false, "message" => "Impossibile salvare il menu"]);
    exit;
  }
}

function resolve_category_key($category) {
  $normalized = strtolower(trim((string)$category));
  if ($normalized === 'drinks' || $normalized === 'drink') {
    return 'drinks';
  }
  if ($normalized === 'foods' || $normalized === 'food') {
    return 'foods';
  }
  return null;
}

function get_item_index($menu, $categoryKey, $name) {
  if (!isset($menu[$categoryKey]) || !is_array($menu[$categoryKey])) {
    return -1;
  }
  for ($i = 0; $i < count($menu[$categoryKey]); $i++) {
    if (isset($menu[$categoryKey][$i]['name']) && $menu[$categoryKey][$i]['name'] === $name) {
      return $i;
    }
  }
  return -1;
}

$menu = read_json_file(MENU_FILE);
$menu = normalize_menu_structure($menu);

if ($method === 'GET') {
  echo json_encode([
    "success" => true,
    "drinks" => $menu['drinks'],
    "foods" => $menu['foods']
  ], JSON_UNESCAPED_UNICODE);
  exit;
}

$actorRole = (string)($payload['actorRole'] ?? '');
if ($actorRole !== 'admin') {
  http_response_code(403);
  echo json_encode(["success" => false, "message" => "Operazione permessa solo agli amministratori"]);
  exit;
}

switch ($method) {
  case 'POST':
    $categoryKey = resolve_category_key($payload['category'] ?? '');
    $name = trim((string)($payload['name'] ?? ''));
    $optionsPayload = $payload['options'] ?? [];
    $options = [];
    if (is_array($optionsPayload)) {
      foreach ($optionsPayload as $option) {
        if (!is_string($option)) continue;
        $trimmed = trim($option);
        if ($trimmed === '' || in_array($trimmed, $options, true)) continue;
        $options[] = $trimmed;
      }
    }

    if ($categoryKey === null) {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Categoria non valida"]);
      exit;
    }
    if ($name === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Il nome della voce è obbligatorio"]);
      exit;
    }
    if (count($options) === 0) {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Specificare almeno una variante"]);
      exit;
    }
    if (get_item_index($menu, $categoryKey, $name) !== -1) {
      http_response_code(409);
      echo json_encode(["success" => false, "message" => "Esiste già una voce con questo nome"]);
      exit;
    }
    $menu[$categoryKey][] = ['name' => $name, 'options' => $options];
    persist_menu($menu);
    echo json_encode(["success" => true]);
    break;

  case 'PUT':
    $categoryKey = resolve_category_key($payload['category'] ?? '');
    $name = trim((string)($payload['name'] ?? ''));
    $updates = is_array($payload['updates'] ?? null) ? $payload['updates'] : [];

    if ($categoryKey === null || $name === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Dati non validi per l'aggiornamento"]);
      exit;
    }

    $index = get_item_index($menu, $categoryKey, $name);
    if ($index === -1) {
      http_response_code(404);
      echo json_encode(["success" => false, "message" => "Voce non trovata"]);
      exit;
    }

    $current = $menu[$categoryKey][$index];
    $newName = trim((string)($updates['newName'] ?? $current['name']));
    if ($newName === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Il nome aggiornato non può essere vuoto"]);
      exit;
    }

    if ($newName !== $current['name'] && get_item_index($menu, $categoryKey, $newName) !== -1) {
      http_response_code(409);
      echo json_encode(["success" => false, "message" => "Esiste già una voce con il nuovo nome"]);
      exit;
    }

    $options = $current['options'];
    if (isset($updates['options'])) {
      $options = [];
      $incoming = $updates['options'];
      if (!is_array($incoming)) {
        http_response_code(400);
        echo json_encode(["success" => false, "message" => "Formato varianti non valido"]);
        exit;
      }
      foreach ($incoming as $option) {
        if (!is_string($option)) continue;
        $trimmed = trim($option);
        if ($trimmed === '' || in_array($trimmed, $options, true)) continue;
        $options[] = $trimmed;
      }
      if (count($options) === 0) {
        http_response_code(400);
        echo json_encode(["success" => false, "message" => "Inserire almeno una variante"]);
        exit;
      }
    }

    $menu[$categoryKey][$index] = ['name' => $newName, 'options' => $options];
    persist_menu($menu);
    echo json_encode(["success" => true]);
    break;

  case 'DELETE':
    $categoryKey = resolve_category_key($payload['category'] ?? '');
    $name = trim((string)($payload['name'] ?? ''));

    if ($categoryKey === null || $name === '') {
      http_response_code(400);
      echo json_encode(["success" => false, "message" => "Dati non validi per l'eliminazione"]);
      exit;
    }

    $index = get_item_index($menu, $categoryKey, $name);
    if ($index === -1) {
      http_response_code(404);
      echo json_encode(["success" => false, "message" => "Voce non trovata"]);
      exit;
    }

    array_splice($menu[$categoryKey], $index, 1);
    persist_menu($menu);
    echo json_encode(["success" => true]);
    break;

  default:
    http_response_code(405);
    echo json_encode(["success" => false, "message" => "Metodo non supportato"]);
}

File: app/routers/menu.py
from __future__ import annotations
from typing import Any, Dict, List

from fastapi import APIRouter
from psycopg.rows import dict_row

from ..common import success, error_response
from ..db import get_connection
from ..utils import normalize_menu_structure, resolve_category_key

router = APIRouter()

@router.get("/menu", response_model=None)
def get_menu():
    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cursor:
            cursor.execute(
                """
                SELECT mi.id, mi.category, mi.name, mo.name AS option_name
                FROM menu_items mi
                LEFT JOIN menu_options mo ON mo.item_id = mi.id
                ORDER BY mi.category, mi.name, mo.name
                """
            )
            rows = cursor.fetchall()

    grouped: Dict[int, Dict[str, Any]] = {}
    for row in rows:
        item_id = row["id"]
        entry = grouped.setdefault(
            item_id,
            {"category": row["category"], "name": row["name"], "options": []},
        )
        option_name = row.get("option_name")
        if option_name and option_name not in entry["options"]:
            entry["options"].append(option_name)

    raw_menu = {"drinks": [], "foods": []}
    for entry in grouped.values():
        raw_menu.setdefault(entry["category"], []).append(
            {"name": entry["name"], "options": entry["options"]}
        )

    normalized = normalize_menu_structure(raw_menu)
    drinks = sorted(normalized["drinks"], key=lambda item: item["name"])
    foods = sorted(normalized["foods"], key=lambda item: item["name"])

    return success(drinks=drinks, foods=foods)

@router.post("/menu", response_model=None)
def add_menu_item(payload: Dict[str, Any]):
    if payload.get("actorRole", "user") != "admin":
        return error_response(403, "Operazione permessa solo agli amministratori")

    category_key = resolve_category_key(payload.get("category"))
    name = str(payload.get("name", "")).strip()
    options_payload = payload.get("options", [])
    options: List[str] = []
    if isinstance(options_payload, list):
        for option in options_payload:
            if isinstance(option, str):
                trimmed = option.strip()
                if trimmed and trimmed not in options:
                    options.append(trimmed)

    if category_key is None:
        return error_response(400, "Categoria non valida")
    if not name:
        return error_response(400, "Il nome della voce è obbligatorio")
    if not options:
        return error_response(400, "Specificare almeno una variante")

    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cursor:
            cursor.execute(
                "SELECT 1 FROM menu_items WHERE category = %s AND name = %s",
                (category_key, name),
            )
            if cursor.fetchone() is not None:
                return error_response(409, "Esiste già una voce con questo nome")

            cursor.execute(
                """
                INSERT INTO menu_items (category, name)
                VALUES (%s, %s)
                RETURNING id
                """,
                (category_key, name),
            )
            item_id = cursor.fetchone()["id"] # type: ignore

            for option in options:
                cursor.execute(
                    "INSERT INTO menu_options (item_id, name) VALUES (%s, %s)",
                    (item_id, option),
                )

            conn.commit()

    return success()

@router.put("/menu", response_model=None)
def update_menu_item(payload: Dict[str, Any]):
    if payload.get("actorRole", "user") != "admin":
        return error_response(403, "Operazione permessa solo agli amministratori")

    category_key = resolve_category_key(payload.get("category"))
    name = str(payload.get("name", "")).strip()
    updates = payload.get("updates", {})

    if category_key is None or not name:
        return error_response(400, "Dati non validi per l'aggiornamento")

    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cursor:
            cursor.execute(
                "SELECT id, name FROM menu_items WHERE category = %s AND name = %s",
                (category_key, name),
            )
            existing = cursor.fetchone()
            if existing is None:
                return error_response(404, "Voce non trovata")

            new_name = str(updates.get("newName", existing["name"])).strip()
            if not new_name:
                return error_response(400, "Il nome aggiornato non può essere vuoto")

            if new_name != existing["name"]:
                cursor.execute(
                    "SELECT 1 FROM menu_items WHERE category = %s AND name = %s",
                    (category_key, new_name),
                )
                if cursor.fetchone() is not None:
                    return error_response(409, "Esiste già una voce con il nuovo nome")

            options = None
            if "options" in updates:
                incoming = updates.get("options")
                if not isinstance(incoming, list):
                    return error_response(400, "Formato varianti non valido")
                parsed: List[str] = []
                for option in incoming:
                    if isinstance(option, str):
                        trimmed = option.strip()
                        if trimmed and trimmed not in parsed:
                            parsed.append(trimmed)
                if not parsed:
                    return error_response(400, "Inserire almeno una variante")
                options = parsed

            if new_name != existing["name"]:
                cursor.execute(
                    "UPDATE menu_items SET name = %s WHERE id = %s",
                    (new_name, existing["id"]),
                )

            if options is not None:
                cursor.execute("DELETE FROM menu_options WHERE item_id = %s", (existing["id"],))
                for option in options:
                    cursor.execute(
                        "INSERT INTO menu_options (item_id, name) VALUES (%s, %s)",
                        (existing["id"], option),
                    )

            conn.commit()

    return success()

@router.delete("/menu", response_model=None)
def delete_menu_item(payload: Dict[str, Any]):
    if payload.get("actorRole", "user") != "admin":
        return error_response(403, "Operazione permessa solo agli amministratori")

    category_key = resolve_category_key(payload.get("category"))
    name = str(payload.get("name", "")).strip()

    if category_key is None or not name:
        return error_response(400, "Dati non validi per l'eliminazione")

    with get_connection() as conn:
        with conn.cursor(row_factory=dict_row) as cursor:
            cursor.execute(
                "SELECT id FROM menu_items WHERE category = %s AND name = %s",
                (category_key, name),
            )
            existing = cursor.fetchone()
            if existing is None:
                return error_response(404, "Voce non trovata")

            cursor.execute("DELETE FROM menu_items WHERE id = %s", (existing["id"],))
            conn.commit()

    return success()